include/mbedtls/rsa.h

```diff
@@ -74,6 +74,169 @@ extern "C" {
 #endif
 
 /**
+ * Helper functions for RSA-related operations on MPI's.
+ */
+
+/**
+ * \brief          Compute RSA prime moduli P, Q from public modulus N=PQ
+ *                 and a pair of private and public key.
+ *
+ * \note           This is a 'static' helper function not operating on
+ *                 an RSA context. Alternative implementations need not
+ *                 overwrite it.
+ *
+ * \param N        RSA modulus N = PQ, with P, Q to be found
+ * \param D        RSA private exponent
+ * \param E        RSA public exponent
+ * \param f_rng    PRNG to be used for randomization, or NULL
+ * \param p_rng    PRNG context for f_rng, or NULL
+ * \param P        Pointer to MPI holding first prime factor of N on success
+ * \param Q        Pointer to MPI holding second prime factor of N on success
+ *
+ * \return         - 0 if successful. In this case, P and Q constitute a
+ *                   factorization of N, and it is guaranteed that D and E
+ *                   are indeed modular inverses modulo P-1 and modulo Q-1.
+ *                   The values of N, D and E are unchanged. It is checked
+ *                   that P, Q are prime if a PRNG is provided.
+ *                 - A non-zero error code otherwise. In this case, the values
+ *                   of N, D, E are undefined.
+ *
+ * \note           The input MPI's are deliberately not declared as constant
```

Comment:
I disagree here and below. If the function fails, no guarantees are made about the arguments, while if it succeeds, it is guaranteed that they haven't been changed. If the user cannot tolerate modification in case of failure, he has to make a copy. On the other hand, making the arguments constant always forces a copy. On the long run, we want to be more space-efficient, and I dislike putting further obstacles in our way by having too many const variables in public, hence not-easy-to-change interfaces.

Reply:
@hanno-arm Several bignum functions including division and gcd already make temporary copies under the hood. Saving one extra bignum in this code is not a major gain. In fact, if you look at what bignum.c is doing under the hood, you may be able to preserve the arguments without making an extra copy — that's an optimization which may or may not be premature at this point.

I maintain my objection against the complexity in the API. It would be legitimate to call these functions with some parameters in read-only memory (e.g. direct read from flash), for example. Modifying input parameters is very surprising behavior, even if the documentation states so. We are making a security library, the API should not be easy to misuse.